iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/CommandDetector.java

```diff
@@ -63,20 +66,69 @@ public static Builder builder() {
    * The foremost element is taken from the stack and checked to see if it matches the command to be searched for.
    */
   public List<Command> search(List<ArgumentResolution> resolvedArguments) {
+    SeparatedList<List<ArgumentResolution>, String> splitCommands = splitCommands(resolvedArguments);
     List<Command> commands = new ArrayList<>();
-    Deque<ArgumentResolution> argumentStack = new LinkedList<>(resolvedArguments);
 
-    PredicateContext context = new PredicateContext(argumentStack, predicates);
+    for (List<ArgumentResolution> resolved : splitCommands.elements()) {
+      Deque<ArgumentResolution> argumentStack = new LinkedList<>(resolved);
 
-    while (!argumentStack.isEmpty()) {
-      List<ArgumentResolution> commandArguments = fullMatch(context);
-      if (!commandArguments.isEmpty()) {
-        commands.add(new Command(commandArguments));
+      PredicateContext context = new PredicateContext(argumentStack, predicates);
+
+      while (!argumentStack.isEmpty()) {
+        List<ArgumentResolution> commandArguments = fullMatch(context);
+        if (!commandArguments.isEmpty()) {
+          commands.add(new Command(commandArguments));
+        }
       }
     }
     return commands;
   }
 
+  private static final String COMMAND_WITHOUT_OPERATOR = "(?:\"(?:\\\\.|[^\"])*+\"|'(?:\\\\.|[^'])*+'|[^;&|])*+";
+  private static final String OPERATOR_REGEX = ";|&&|&|\\|\\||\\|";
+  private static final Pattern PATTERN_MULTIPLE_OPERATOR_DETECTION_BEFORE = Pattern.compile(
+    "(?<before>" + COMMAND_WITHOUT_OPERATOR + ")(?<operator>" + OPERATOR_REGEX + ")");
+
+  // Cognitive Complexity of methods should not be too high
+  @SuppressWarnings("java:S3776")
+  private static SeparatedList<List<ArgumentResolution>, String> splitCommands(List<ArgumentResolution> resolvedArguments) {
+    List<List<ArgumentResolution>> listOfArgumentList = new ArrayList<>();
```

Comment:
This would have some advantages, as you mentioned for example the possibility to reuse the existing `SeparatedList` implementation.
But I believe we would lose important structure information and would be simply an equivalent of stripping off any multi-command operator. It would result in the exact same behavior for these two very different instructions:
```
command1; command2
command1 command2
```

Reply:
We don't use separators right now, but I can imagine that they can be useful in the future.

Reply:
I looked again at the implementation of `ArgumentResolution`, you're right, it makes sense to keep a list of lists.